doc/conventions.md

```diff
@@ -0,0 +1,208 @@
+Providing components provide values of type `*Endpoint`.
+
+The collecting component should define the registration type and a constructor for it.
+
+```go
+// --- server/component.go ---
+
+// Endpoint is provided by components that wish to register an endpoint
+// with this server.  If a nil *Endpoint is provided, it will be ignored.
+type Endpoint struct { .. }
+
+// NewEndpoint creates a new Endpoint.
+func NewEndpoint(route string, handler func()) *Endpoint { .. }
+```
+
+Its implementation then requires a slice of the registration type, using `group:"true"`:
+
+```go
+// --- server/server.go ---
+type dependencies struct {
+    fx.In
+
+    Endpoints []*Endpoint `group:"true"`
+}
+
+func newServer(deps dependencies) Component {
+    // ..
+    for _, e := range deps.Endpoints {
+        if e == nil {
+            continue
+        }
+        // ..
+    }
+    // ..
+}
+```
+
+It's good practice to ignore nil values, as that allows providing components to skip
+the registration if desired.
+
+Finally, the providing component includes a registration in its output:
+
+```go
+// --- foo/foo.go ---
+type provides struct {
+    fx.Out
+
+    Component
+    ServerRegistration *server.Registration `group:"true"`
+}
+
+func newFoo(deps dependencies) provides {
+    // ..
+    return provides{
+        Component: foo,
+        ServerRegistration: server.NewRegistration("/things/foo", foo.handler),
+    }
+}
+```
+
+This technique has some caveats to be aware of:
+
+ * The providing components are instantiated before the collecting component.
+ * Fx treats value groups as the collecting component depending on all of the providing components.
+   This means that the providing components cannot depend on the collecting component.
+ * Fx will instantiate _all_ providing components before the collecting component.
+   This may lead to components being instantiated in unexpected circumstances.
+   The AutoStart convention is meant to partially address this issue.
+ * Omitting the `group:"true"` in either place it appears above will lead to Fx silently ignoring the registration.
+
+## Subscriptions
+
+Subscriptions are a common form of registration, and have support in the `pkg/util/subscriptions` package.
+
+To implement a subscription, the collecting component defines a message type, a Subscription type, and a Subscribe function:
+
+```go
+// --- eventpub/component.go ---
+type Event struct { .. }
+
+// Subscription is provided by components that wish to subscribe to events
+// from this component.  A nil *Subscription will be ignored.
+type Subscription = subscriptions.Subscription[Event]
+
+// Subscribe creates a new Subscription.
+func Subscribe() Subscription {
+    return subscriptions.NewSubscription[Event]()
+}
+
+// --- eventpub/eventpub.go ---
+type eventpub struct {
+    // .. (*eventpub implements Component)
+    subscriptions *subscriptions.SubscriptionPoint[Event]
+}
+
+type dependencies struct {
+	fx.In
+
+    // ...
```

Comment:
```suggestion
	// ...
```

Reply:
Ugh, Go and its Tabs.  1970 called and wants its editor back.